### DISCUSSION TITLE
How to link a MOOSE App with Blackbear and NEML?

### DISCUSSION THREAD

**Discussion by ngrilli** [idaholab/blackbear#337] [General]
Dear Blackbear users,

I am trying to link a MOOSE App including some custom code with blackbear, specifically I want to use NEML objects,
I have made different attempts and the compile goes well but it always gives the error saying the NEMLStress object
is not registered.

I tried the following:

Installing blackbear, then NEML with the standard procedure
and everything works well.

Then I tried both:

1) Creating a MOOSE App using the standard procedure here:
https://mooseframework.inl.gov/getting_started/new_users.html
this attempt was made using my existing moose installation,
which should be a newer version compared with the moose installation inside blackbear.

2) Same but I create the App based on the script stork.sh
that is contained inside the blackbear folder, meaning inside the moose version
that is in the blackbear folder.

Once I create my App, I change the Makefile to include all the modules:

ALL_MODULES                 := yes

I also tried with or without adding a link to blackbear in the Makefile as follows:

BLACKBEAR_DIR      ?= $(CURDIR)/../blackbear
APPLICATION_DIR    := $(BLACKBEAR_DIR)
APPLICATION_NAME   := blackbear
include            $(FRAMEWORK_DIR)/app.mk

My App compiles correctly.
However, when I pick any example input file in the blackbear folder
in which a NEML object is used, and I try to run it with the executable
created in my App, I get for instance:

"NEMLStress object is not registered"

How do you normally link a custom App to blackbear and NEML?

Thank you very much in advance for your help,
Nicolò

**Comment by bwspenc** [comment]
@ngrilli It's definitely possible to do this because we do exactly what you're trying to do with the Grizzly application, which pulls in BlackBear. I vaguely recall having a similar issue when we originally set that up. I have to leave for the day right now, but I'll look into what we did and get back to you tomorrow.

**Comment by bwspenc** [comment]
I think what you have in your Makefile is correct, and that all you're missing is a couple of lines in your AppNameApp.C file (where AppName is the name of your app):
```
#include "BlackBearApp.h"
```
Also, in your `AppNameApp::registerAll()` method, alongside your call to `ModulesApp::registerAll()`, you also need to add a call to that same method for BlackBearApp:
```
  BlackBearApp::registerAll(factory, action_factory, syntax);
```
